### PR TITLE
enable VPSM4_EX_CAPABLE for sm4_gcm

### PR DESCRIPTION
**Pull request by zzl360**
type     16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
before:
SM4-GCM  71134.38k    81512.39k    83432.11k    83808.60k    85183.15k    86199.57k
after:
SM4-GCM  30656.22k   108801.82k   196005.97k   198715.73k   199349.59k   199387.82k

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
There is quite severe performance degradation for the 16 bytes case. Could it be somehow avoided/minimized?

**Comment by zzl360** [comment]
> There is quite severe performance degradation for the 16 bytes case. Could it be somehow avoided/minimized?

this is explained by https://github.com/openssl/openssl/pull/19914#issuecomment-1358958883

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by zorrorffm** [comment]
To verify the impact on platform supporting HWSM4_CAPABLE, I will have a performance test on N2 platform.

**Comment by zorrorffm** [comment]
Performance test shows no performance regression on other platforms(N1, N2).  So looks good to our side.

**Comment by zzl360** [comment]
> Performance test shows no performance regression on other platforms(N1, N2). So looks good to our side.

yes,this patch only impact KunPeng920 platform.

**Comment by t8m** [comment]
Merged to master branch. Thank you for your contribution.

File: providers/implementations/ciphers/cipher_sm4_gcm_hw.c [hunk 1]
```diff
@@ -14,37 +14,46 @@
...
 #  else /* HWSM4_ctr32_encrypt_blocks */
-        ctx->ctr = (ctr128_f)NULL;
+        SM4_GCM_HW_SET_KEY_CTR_FN(ks, HWSM4_set_encrypt_key, HWSM4_encrypt, NULL);
 #  endif
     } else
 # endif /* HWSM4_CAPABLE */
+
+#ifdef VPSM4_EX_CAPABLE
+    if (VPSM4_EX_CAPABLE) {
+        SM4_GCM_HW_SET_KEY_CTR_FN(ks, vpsm4_ex_set_decrypt_key, vpsm4_ex_encrypt,
```
Comment (xu-yi-zhou) [inline providers/implementations/ciphers/cipher_sm4_gcm_hw.c:43]: @zzl360 sm4-gcm test failures when I run `make TESTS=test_evp test` on Kunpeng-920, maybe the `fn_set_enc_key` should be `vpsm4_ex_set_encrypt_key` here?

Comment (zzl360) [reply providers/implementations/ciphers/cipher_sm4_gcm_hw.c:43]: yes, I'll fix it soon.

Comment (zzl360) [reply providers/implementations/ciphers/cipher_sm4_gcm_hw.c:43]: > yes, I'll fix it soon.

#20751

**Pull request closed without merging**